x-pack/filebeat/input/awss3/interfaces.go

```diff
@@ -285,6 +287,17 @@ func (a *awsS3API) DeleteObject(ctx context.Context, bucket, key string) (*s3.De
 	return deleteObjectOutput, nil
 }
 
+func (a *awsS3API) clientFor(region string) *s3.Client {
+	// Conditionally replace the client if the region of
+	// the request does not match the pre-prepared client.
+	opts := a.client.Options()
+	if opts.Region == region {
+		return a.client
+	}
+	opts.Region = region
+	return s3.New(opts)
```

Comment:
Depending on the frequency of the new-client path, this could generate a reasonable amount of work. Something that I considered but did not add is having `awsS3API` hold a `map[string]*s3.Client` that can cache (non-retiring) the clients for each region. Given that there are not large numbers of regions, it should be reasonable to retain clients that are constructed dynamically for the life of the `awsS3API`. WDYT?

Reply:
I think this is a good idea. My intuition is that in the use-cases where this is needed that it can be in very high volumes (big orgs, lots of aws projects, lots of regions, all notifications centralized to one queue).

Reply:
I've added caching and limited the size to 100. We should never hit this limit, but I wanted to make sure that we don't put ourselves in the situation that this results in an OoM crash.